controllers/operatorpolicy_controller.go

```diff
@@ -133,9 +146,96 @@ func (r *OperatorPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 
 	// FUTURE: more resource checks
 
+	// handle catalogsource
+	if err := r.handleCatalogSource(ctx, policy); err != nil {
+		OpLog.Error(err, "Error handling CatalogSource")
+
+		return reconcile.Result{}, err
+	}
+
 	return reconcile.Result{}, nil
 }
 
+func (r *OperatorPolicyReconciler) handleCatalogSource(
+	ctx context.Context,
+	policy *policyv1beta1.OperatorPolicy,
+) error {
+	watcher := opPolIdentifier(policy.Namespace, policy.Name)
+
+	desiredSub, err := buildSubscription(policy)
+	if err != nil {
+		return fmt.Errorf("error building Subscription: %w", err)
+	}
+
+	// Retrieve Subscription
+	foundSub, err := r.DynamicWatcher.Get(watcher, subscriptionGVK, desiredSub.Namespace, desiredSub.Name)
+	if err != nil {
+		return fmt.Errorf("error getting Subscription: %w", err)
+	}
+
+	if foundSub == nil {
+		// Subscription not found, check the CatalogSource object
+		foundCatalogSrc, err := r.DynamicWatcher.Get(watcher, catalogSrcGVK,
+			desiredSub.Spec.CatalogSourceNamespace, desiredSub.Spec.CatalogSource)
+		if err != nil {
+			return fmt.Errorf("error getting CatalogSource: %w", err)
+		}
+
+		// Conditionally emit health event using the defined status template
+		isMissing := foundCatalogSrc == nil
+
+		err = r.updateStatus(ctx, policy, catalogSourceFindCond(isMissing), catalogSourceObj(desiredSub, isMissing))
+		if err != nil {
+			return fmt.Errorf("error updating the status for CatalogSource when Subscription DNE on cluster: %w", err)
+		}
+
+		return nil
+	}
+
+	condList, found, err := unstructured.NestedSlice(foundSub.Object, "status", "conditions")
+	if !found {
+		return fmt.Errorf("field does not exist in object: %w", err)
+	}
+
+	if err != nil {
+		return fmt.Errorf("error decoding object into array type: %w", err)
+	}
+
+	// Iterate over all the conditions to determine the state of the CatalogSource
+	for _, condition := range condList {
+		condMap, ok := condition.(map[string]interface{})
+		if !ok {
+			return fmt.Errorf("conditions field is of the wrong type")
+		}
+
+		if fmt.Sprintf("%v", condMap["type"]) == catalogSrcConditionType {
+			if fmt.Sprintf("%v", condMap["status"]) == string(metav1.ConditionFalse) {
+				// Emit healthy event using the defined status template
+				err := r.updateStatus(ctx, policy, catalogSourceFindCond(false), catalogSourceObj(desiredSub, false))
+				if err != nil {
+					return fmt.Errorf("error updating the status for an existing CatalogSource: %w", err)
+				}
+			} else {
+				// Emit unhealthy event by inheriting condition fields from the Subscription
+				failureCond := metav1.Condition{
+					Type:               fmt.Sprintf("%v", condMap["type"]),
+					Status:             metav1.ConditionTrue,
+					LastTransitionTime: metav1.Now(),
```

Comment:
Is this not in the condition on the subscription?

Reply:
@JustinKuli

Reply:
I think you would need to marshal it to json and then unmarshal it again, or maybe you can use an unstructured converter.

I think a better approach here overall would be to convert the Subscription itself out of Unstructured, and then you get all the conditions and other details more easily.